### ISSUE TITLE
Indent guides should not extend beyond the last line

### ISSUE THREAD

**Issue by aspizu** [helix-editor/helix#8487]
Indent guides should stop at the last non-empty line inside an indented block in languages which use indentation for structure.

left: what helix does currently.
right: what helix should do.
![Issue demonstrated in image.](https://media.discordapp.net/attachments/972556399928299661/1160572370508062780/image.png?ex=6535263f&is=6522b13f&hm=615c284d0dcc399027878642478bba736a3b54198d63190b28313f84e59bac55&=&width=1158&height=685)

**Comment by aspizu** [comment]
So, helix should add a option which will toggle this behavior and depending on the file-type enable this behavior. Languages such as Python would enable this. It could be enabled for any file optionally.

How difficult would be to implement this?

**Comment by kirawi** [comment]
https://github.com/helix-editor/helix/blob/a8574805617d86304232b47d84fd7c7f4f773520/helix-term/src/ui/document.rs#L154

**Comment by aspizu** [comment]
Could tree-sitter be used for determining where to draw them?

**Comment by pascalkuthe** [comment]
I don't think we should start using TS for indent guides.

I think the specific of how we render them is simply not ideal.

We continue the indentation of the previous line across emply and whitespace only line.

Instead we should:
* use the trailing indentation for empty lines (fixing this issue)
* compute the indentation for whitespacelines as normal. This makes sure that `o` at the end of a function still shows the correct indent guides

This is how other editors/vim plugins handle this. That approach would also allow us to do some lookahead to avoid indent guides disappearing at the first/last lines as they currently do.

**Comment by aspizu** [comment]
Code formatters remove trailing indentation which would introduce another issue.

Using TS for indent guides would also allow us to fix this mess.
![](https://media.discordapp.net/attachments/448238009733742612/1161249924491526234/image.png?ex=65379d44&is=65252844&hm=f1071397c357abe768eed892e571cf7e7cc1e32dabac9132106e859cf668d480&=&width=1086&height=204)

**Comment by pascalkuthe** [comment]
yes that is exactly the intention. Helix also removes trailing indentation when leaving insert mode (if the line is empty on the cursor line). This ensures that the indent guides both behave correctly and use the trailing indentation (the indentatino of the next non-emtpy line) while also making sure that lines which are manually indented but don't contain text yet (lime after hitting `o`) have the correct indentation.

I don't think the case you show should be fixed. Indentguides are not meant to show somekind of "language level scope& or similar. They are supped to show how many times you need to press tab to reach te same indentatino. They are purely textual and I don't think that should change.

**Comment by kirawi** [comment]
> use the trailing indentation for empty lines (fixing this issue)

What does "trailing indentation" mean in this context?

**Comment by pascalkuthe** [comment]
The indentation of the first non-empty line after an empty lines
